### ISSUE TITLE
LED strobing

### ISSUE THREAD

**Issue by alexbbrown** [rogerlinndesign/linnstrument-firmware#14]
I note in videos that the LEDs pulsate.  

This is most likely due to aliasing artefacts with the video recording hardware (often 25 or 30 or 60fps).
You might be getting pulsing due to your low refresh rate.
Or there might be pulsing due to delays affecting the light timing changing at visible frequencies.

I can't check yet; my linnstrument is wrapped up for xmas, so I'm going to try to understand this code instead – so sorry for this rambling post :). 

Your LED refresh rate is at 500us, which covers 26 light cycles (assuming there's no pure off cycle apart from the power saving one).  This leads to an on-time of 13:1000, each light is lit for up to ~1/77s  (77fps).  This frequency is not in the visible range, but the beat frequency with a 60 fps modern video camera is 17fps, which is visible. A PAL derived camera at 50 might be visible at 27fps  . 

This can be fixed by pushing the beat frequency up to 30 or so - try setting the delay to 430 instead of 500.  If this fixes it, then rock'n'roll videos featuring the linnstrument will be much improved.  

Note: do you get visible shearing or bending of text on video recordings of the linnstrument when scrolling messages?  There may be a different fix for that, but to start try reducing the interval still further.  If you do you'll need to uncouple the scroll rate (and pedal sampling) from the update rate.

---

What following is a suggestion for addressing jitter in the delay – which is implied by your code but may not have visual effect beyond the above aliasing with PAL.  Do the lights pulsate or flash bright and dim visibly to the eye?  if so, read on.

Note: You don't only update LEDs on this schedule, but also check the foot pedals and do scrolling.  I suspect timing defects (phase shifts) in scrolling rate are highly visible, even if a momentary brightness is unimportant, it may cause other problems.  Does the delay - see below - every get worse when playing heavily?  or when getting a lot of traffic over midi?

----

Right now, the code adjusts the LEDs on and off at time T + d, where T was the last time you updated a column, and d1 is the (unintended) delay - how late this refresh is.  You then set T' to T + d, and switch the LEDs.  Every value of d is different, and is always zero or positive.

I would be interested to know if you have measured the variance in d (range or better interquartile range t1 should suffice).

This way of accounting for delay means that that variations in the value of d over time will cause LED brightness to vary, and are cumulative, which can cause scrolling to look weird.  The cumulative effects correct over time (fractions of a second), since the phase shifts, and the new phase will become the new norm.

The variance in the delay has two impacts on brightness.  1) leaving a LED lit for (e.g) 2x longer than usual will transiently make lit LEDs appear twice as bright (for one frame of 1/80 second).  This might not be visible, but might have a 'sparkling' effect.  It will also leave all other other lights OFF for 1/2000 of a second longer, and reduce their average brightness for the second by 1/26 (about 4%).  It's unclear to me if the resulting reduction in light intensity will be visible, unless the delay has its own intrinsic frequency in the visible range.

The current code allows the scheduling slot to drift (adding the delay each time) - the gain and lit time is never reclaimed.  This means if the amount by which the refresh is late is positive, large, high variance and varies at visible frequencies (I guess ~10Hz), the leds will visibly change in brightness over time.

A simple fix would be to calculate the next slice time relative to the expected time, instead of now:

inline boolean checkRefreshLedColumn(unsigned long now) {
  if (calcTimeDelta(now, prevLedTimerCount) > ledRefreshInterval) {        // is it time to refresh the next LED column?
    refreshLedColumn(now);                                                 // yes, refresh the next LED column...
//    prevLedTimerCount = now;                                               // and reset the LED timer count to current    
    prevLedTimerCount += ledRefreshInterval;                                               // and reset the LED timer count to the time we expected to update it
    return true;
  }
  return false;
}

this might instead be rewritten using a variable to hold the next expected time.

The above fix will perform less than perfectly because not all columns are updated at once.  The effect will one bright column followed by a darker column for 1 microsecond.  Unless you have periodicity in the delay at frequencies related to the update frequency you are probably OK.

----

It can be improved by storing the delay (lateness) measured at each interval, and crediting that to the column in question (the column we are about to update).  The next time we are scheduling the previous (in time) column, adjust the 'next expected time' to add the deficit.

In other words, if a column is lit for 500 * 1.5 = 750ms in time slice 100, then in time slice 126 it will be lit for 500 - 250 = 250, so the average brightness over the two slices is still 500.

static nextColumnToUpdate

----

Hope these suggestions are useful.

**Comment by gbevin** [comment]
Thanks for the suggestions, setting it to 430 does nothing in terms of pulsating lights. Most cameras nowadays are neither PAL nor NTSC and record at much higher frame rates.

The reason why the LEDs are backed off as much is that if we given them more importance, it has a noticeable impact on playing latency and jitter. After many months of tuning we finally came to a system where the playing latency is amazingly low, with the LEDs looking uniform to the naked eye.

The only way to get rid of pulsating LEDs on videos is to either use a more powerful Arduino (which doesn't exist yet), or to introduce a dedicated LED refresh processor, for which it was too late in the hardware design process without imposing significant delays to the initial release.

**Comment by alexbbrown** [comment]
Thanks for looking at it, I'm having fun reading the code anyway.
